File: web/src/features/watch/application/narration-vtt.ts
/**
 * WebVTT → narration cues.
 *
 * Pure: a string in, cues out. It was previously welded to a `fetch` call, which
 * meant the only way to try it on a real caption file was to run the app and
 * listen — and listening is how "sometimes a line repeats, sometimes one goes
 * missing" stayed unresolved. Separated, it can be run against the actual files
 * on disk.
 *
 * Two caption shapes arrive from YouTube and they are not variations of one
 * format, they are different formats:
 *
 *   manual    one cue, one or more plain lines, complete sentences already
 *   automatic rolling cues, where each carries the tail of the previous line
 *             plus one new phrase, the new phrase timestamped word by word
 *
 * The automatic shape is the awkward one. Its cues overlap by design, and the
 * per-word timestamps are the only accurate timing in the file.
 *
 * That overlap is handled where it arises — by reading only the new line of a
 * rolling block — and nowhere else. There used to be a pass at the end that
 * compared each cue with the one before it and stripped whatever prefix they
 * shared, if the shared part covered half the earlier cue. It could not tell a
 * rolling repeat from two lines that genuinely begin alike, so it deleted real
 * captions as readily as duplicated ones: one guess producing both the
 * complaints it was meant to prevent. A file where two speakers each say "Come
 * on." loses the second of them to a rule like that. Removing it costs nothing,
 * because reading only the new line leaves nothing for it to find.
 */

export interface CueText {
  start: number
  end: number
  text: string
}

/** WebVTT timestamp → seconds. "00:01:23.456" → 83.456 */
export function parseVTTTime(raw: string): number {
  const parts = raw.split(':')
  if (parts.length !== 3) return NaN
  const h = Number.parseInt(parts[0], 10)
  const m = Number.parseInt(parts[1], 10)
  const s = Number.parseFloat(parts[2])
  return h * 3600 + m * 60 + s
}

/**
 * Strip WebVTT markup and the symbols a speech synthesiser would read out.
 *
 * `[cười]`, `[thở dài]`, `[hắng giọng]` survive: the TTS server understands
 * those as emotion. Other bracketed descriptions are removed by stripBrackets,
 * later, because a bracket can span several `<c>` tags and would not be whole
 * at this point.
 */
export function cleanCueText(raw: string): string {
  let s = raw
    // Entities first, so &gt;&gt; becomes >> in time to be recognised below.
    .replace(/&amp;/g, '&')
    .replace(/&lt;/g, '<')
    .replace(/&gt;/g, '>')
    .replace(/&quot;/g, '"')
    .replace(/&#39;/g, "'")

  // The four quote marks are written as escapes on purpose. As literals they
  // were at some point flattened to straight " and ' — by an editor or a
  // formatter, silently — and the class then stripped the apostrophe out of
  // every contraction it saw: "they're" reached the translator as "theyre".
  s = s
    .replace(/<[^>]+>/g, '')
    .replace(/>>\s*/g, '')
    .replace(/[♪♫♬→←↑↓↔«»\u201C\u201D\u2018\u2019„‚]/g, '')
    .trim()

  return s.replace(/\s{2,}/g, ' ')
}

/** Remove [sound effects] but keep the emotion tags the TTS understands. */
export function stripBrackets(text: string): string {
  const EMOTION_TAGS = /\[(cười|thở dài|hắng giọng)\]/gi
  const placeholders: string[] = []
  let s = text.replace(EMOTION_TAGS, (match) => {
    placeholders.push(match)
    return `\x00${placeholders.length - 1}\x00`
  })
  s = s.replace(/\s*\[[^\]]*\]\s*/g, ' ')
  s = s.replace(/ \./g, '.')
  s = s.replace(/\x00(\d+)\x00/g, (_, i) => placeholders[+i] || '')
  s = s.replace(/\s{2,}/g, ' ')
  return s.trim()
}

/** Cues shorter than this are YouTube's clean-snapshot copies, not speech. */
const SNAPSHOT_MAX_SECONDS = 0.1

/** The clause a comma opens is not split off if it is shorter than this. */
const MIN_CLAUSE_WORDS = 3

/**
 * Nor is the clause a comma closes.
 *
 * Larger than MIN_CLAUSE_WORDS, and measured rather than chosen. The rule above
 * exists to stop a clause being voiced as a clip half a second long — and at 3
 * it was not stopping it: on one measured video thirteen clips came out under a
 * second and the shortest, "And of course,", was 0.40s. Every one of them was an
 * adverbial opener that had picked up three words and qualified.
 *
 * English ASR here runs at 3.37 words a second, so five words is about a second
 * and a half. At five, those thirteen become three — and the three that remain
 * are whole short sentences, which are allowed to be short.
 */
const MIN_CLAUSE_BEFORE = 5

/**
 * With no punctuation at all, a clause is cut once it reaches this length.
 *
 * A ceiling on speech that never punctuates, and nothing else: any boundary at
 * all is taken first, on the word it arrives. It was 30, which is low enough to
 * fire a word or two *before* a full stop — measured, it cut "...or maybe okay
 * maybe slightly" and left "below those." as a cue of its own, 0.72s long. A
 * blind cut that beats a real boundary to it by two words is the one thing this
 * constant must not do, so it has room to let a sentence finish.
 */
const FORCE_SPLIT_WORDS = 40

const ABBREVIATIONS = /^(Dr|Mr|Mrs|Ms|DR|Prof|Sr|Jr|vs|etc)$/i

/** One word or phrase with the time it is spoken. */
interface Piece {
  start: number
  end: number
  text: string
}

interface RawBlock {
  start: number
  end: number
  lines: string[]
}

/**
 * Split the file into timed blocks.
 *
 * The payload of a block ends at a blank line *or* at the next timing line —
 * files in the wild have both. Only the blank separator may be stepped over;
 * stepping over a timing line, which the previous version did unconditionally,
 * discards the whole cue that line introduced.
 */
function readBlocks(raw: string): RawBlock[] {
  const lines = raw.replace(/\r\n?/g, '\n').split('\n')
  const blocks: RawBlock[] = []
  let i = 0

  while (i < lines.length) {
    if (!lines[i]?.includes('-->')) {
      i++
      continue
    }

    const timing = lines[i]
    const arrow = timing.indexOf('-->')
    const start = parseVTTTime(timing.slice(0, arrow).trim())
    const end = parseVTTTime(timing.slice(arrow + 3).trim().split(/\s/)[0])
    i++

    // Leading blank lines belong to the separator, not to the payload.
    while (i < lines.length && lines[i].trim() === '') i++

    const payload: string[] = []
    while (i < lines.length && lines[i].trim() !== '' && !lines[i].includes('-->')) {
      payload.push(lines[i])
      i++
    }

    if (Number.isFinite(start) && Number.isFinite(end)) {
      blocks.push({ start, end, lines: payload })
    }
  }

  return blocks
}

/**
 * Turn one automatic-caption block into word-level pieces.
 *
 * Only the last payload line is read. In the rolling format the earlier lines
 * are the previous cue's text, repeated for the viewer's benefit — reading them
 * is what makes a phrase say itself twice.
 */
function piecesFromTagged(block: RawBlock): Piece[] {
  const line = block.lines[block.lines.length - 1] ?? ''
  const pieces: Piece[] = []

  // Text before the first timestamp tag is spoken from the cue's own start.
  const lead = line.match(/^([^<]+)</)
  if (lead) {
    const text = cleanCueText(lead[1])
    if (text) pieces.push({ start: block.start, end: 0, text })
  }

  const wordRe = /(\d{2}:\d{2}:\d{2}\.\d{3})><c>([^<]*)<\/c>/g
  let m: RegExpExecArray | null
  while ((m = wordRe.exec(line)) !== null) {
    const at = parseVTTTime(m[1])
    const text = cleanCueText(m[2])
    if (text && Number.isFinite(at)) pieces.push({ start: at, end: 0, text })
  }

  if (pieces.length === 0) {
    const text = cleanCueText(line)
    if (text) pieces.push({ start: block.start, end: block.end, text })
    return pieces
  }

  // Each word runs until the next one begins; the last runs to the cue's end.
  for (let k = 0; k < pieces.length; k++) {
    pieces[k].end = k + 1 < pieces.length ? pieces[k + 1].start : block.end
  }
  return pieces
}

/**
 * Where the first clause boundary in `text` is, or -1.
 *
 * The *first*, not the last. Taking the last swallowed every boundary before it,
 * so "Hello there, my friend. How are you" came out as a single cue instead of
 * the three the rule asks for.
 */
/** The text from just past `idx` to the next punctuation mark, or to the end. */
function segmentAfter(text: string, idx: number): string {
  const next = text.slice(idx + 1).search(/[.!?,]/)
  return next < 0 ? text.slice(idx + 1) : text.slice(idx + 1, idx + 1 + next)
}

export function firstClauseBoundary(text: string): number {
  const re = /[.!?,]/g
  let m: RegExpExecArray | null
  while ((m = re.exec(text)) !== null) {
    const ch = m[0]
    const idx = m.index
    const before = text[idx - 1]
    const after = text[idx + 1]

    // "2.5", "3.14" — a decimal point is not the end of anything.
    if (ch === '.' && before && /\d/.test(before) && after && /\d/.test(after)) continue

    if (ch === '.' && before && /[A-Za-z]/.test(before) && after === ' ') {
      const wordBefore = text.slice(0, idx).split(/\s+/).pop() ?? ''
      if (ABBREVIATIONS.test(wordBefore)) continue
    }

    // Mid-word punctuation is not a boundary.
    if (after && after !== ' ') continue

    // A comma is only a boundary when both sides can stand alone. Splitting
    // "Then, how are you" leaves "Then," to be translated with no sentence
    // around it and voiced as a clip half a second long — the opposite of what
    // splitting is for.
    //
    // The side that follows is the *next clause*, not everything left in the
    // buffer. Measured to the end it counted the whole rest of the sentence, so
    // a list separator always qualified: "the AirPods 4, 3, or 2" was cut after
    // "4," and left "3, or 2." to be spoken alone, 0.72s long.
    if (ch === ',') {
      const wordsBefore = text.slice(0, idx).trim().split(/\s+/).filter(Boolean).length
      const wordsAfter = segmentAfter(text, idx).trim().split(/\s+/).filter(Boolean).length
      if (wordsBefore < MIN_CLAUSE_BEFORE || wordsAfter < MIN_CLAUSE_WORDS) continue
    }

    return idx + 1
  }
  return -1
}

/**
 * Gather word pieces into clauses.
 *
 * A clause carries the start of its first word and the end of its last, which
 * is the whole point of keeping per-word timing this far: a clause that claimed
 * the timing of the block it came from would be wrong by however much of that
 * block belonged to the clause before it.
 */
function groupIntoClauses(pieces: Piece[]): CueText[] {
  const out: CueText[] = []
  let buf: Piece[] = []

  const emit = (upTo: number) => {
    // upTo is a character offset into the joined text.
    let consumed = 0
    let cut = -1
    let leftover: Piece | null = null

    for (let k = 0; k < buf.length; k++) {
      const next = consumed + (k > 0 ? 1 : 0) + buf[k].text.length
      if (upTo <= next) {
        cut = k
        const within = upTo - (consumed + (k > 0 ? 1 : 0))
        const rest = buf[k].text.slice(within).trim()
        if (rest) leftover = { start: buf[k].start, end: buf[k].end, text: rest }
        break
      }
      consumed = next
    }
    if (cut < 0) cut = buf.length - 1

    const taken = buf.slice(0, cut + 1)
    const text = stripBrackets(taken.map((p) => p.text).join(' ').slice(0, upTo).trim())
    if (text) {
      out.push({ start: taken[0].start, end: taken[taken.length - 1].end, text })
    }
    buf = leftover ? [leftover, ...buf.slice(cut + 1)] : buf.slice(cut + 1)
  }

  for (const piece of pieces) {
    // A piece that is nothing but a bracketed description is a timestamp with
    // no speech under it. Left in the buffer it is stripped at the end of emit,
    // by which time the clause has already taken its start — so the line after
    // it is spoken from the moment the music began.
    if (!stripBrackets(piece.text)) continue
    buf.push(piece)

    // Loop: one block can complete more than one clause.
    for (;;) {
      const joined = buf.map((p) => p.text).join(' ')
      const at = firstClauseBoundary(joined)
      if (at < 0) break
      emit(at)
      if (buf.length === 0) break
    }

    const joined = buf.map((p) => p.text).join(' ')
    if (joined.trim().split(/\s+/).filter(Boolean).length >= FORCE_SPLIT_WORDS) {
      emit(joined.length)
    }
  }

  if (buf.length > 0) {
    const text = stripBrackets(buf.map((p) => p.text).join(' ').trim())
    if (text) out.push({ start: buf[0].start, end: buf[buf.length - 1].end, text })
  }

  return out
}

/**
 * Parse a WebVTT file into the cues the narration speaks.
 *
 * `sourceLang` decides whether clause splitting applies: it is tuned for the
 * punctuation of English and Vietnamese, and a language it does not know is
 * better left in whole cues than cut on rules that do not hold there.
 */
export function parseVTT(raw: string, sourceLang = 'vi'): CueText[] {
  const blocks = readBlocks(raw)
  const isAuto = blocks.some((b) => b.lines.some((l) => l.includes('<c>')))

  const pieces: Piece[] = []
  const manual: CueText[] = []

  for (const block of blocks) {
    // YouTube's ~10ms clean copies of a line already carried by a real cue.
    if (block.end - block.start < SNAPSHOT_MAX_SECONDS) continue
    if (block.lines.length === 0) continue

    const tagged = block.lines[block.lines.length - 1]?.includes('<c>') ?? false

    if (tagged) {
      pieces.push(...piecesFromTagged(block))
      continue
    }

    if (isAuto) {
      // Rolling text without tags: again only the last line is new.
      const text = cleanCueText(block.lines[block.lines.length - 1] ?? '')
      if (text) pieces.push({ start: block.start, end: block.end, text })
      continue
    }

    const text = cleanCueText(block.lines.join(' '))
    if (text) manual.push({ start: block.start, end: block.end, text })
  }

  if (!isAuto) {
    return manual.map((c) => ({ ...c, text: stripBrackets(c.text) })).filter((c) => c.text)
  }

  if (sourceLang !== 'en' && sourceLang !== 'vi') {
    return pieces
      .map((p) => ({ start: p.start, end: p.end, text: stripBrackets(p.text) }))
      .filter((c) => c.text)
  }

  return groupIntoClauses(pieces)
}

File: services/gateway/internal/api/narration_vtt.go
package api

import (
	"math"
	"regexp"
	"strconv"
	"strings"
)

// WebVTT -> narration cues.
//
// A port of web/src/features/watch/application/narration-vtt.ts, moved here so
// the narration pass can run on the server. The reasoning in that file holds
// unchanged and is not repeated; what follows records only what is different in
// Go, and the one thing the port had to get right that TypeScript got for free.
//
// **Offsets are runes, not bytes.** The original counts in UTF-16 code units and
// slices the joined text by those offsets. Counting bytes here would put every
// cut in the wrong place the moment a line contains a Vietnamese vowel — and the
// source of these captions is a library whose whole point is that it holds
// Vietnamese. Every offset below is an index into a []rune.
//
// Two caption shapes arrive from YouTube and they are different formats, not
// variations of one: manual cues carry complete sentences, automatic cues roll,
// each carrying the tail of the previous line plus one new phrase timestamped
// word by word.

// vttCue is one line of narration with the seconds it occupies.
type vttCue struct {
	Start float64 `json:"startSeconds"`
	End   float64 `json:"endSeconds"`
	Text  string  `json:"text"`
}

// snapshotMaxSeconds: cues shorter than this are YouTube's clean-snapshot
// copies of a line a real cue already carries, not speech.
const snapshotMaxSeconds = 0.1

// minClauseWords: a clause is not split off a comma if either side is shorter.
const minClauseWords = 3

// forceSplitWords: with no punctuation at all, a clause is cut at this length.
const forceSplitWords = 30

var (
	abbreviations = regexp.MustCompile(`(?i)^(Dr|Mr|Mrs|Ms|Prof|Sr|Jr|vs|etc)$`)
	tagRe         = regexp.MustCompile(`<[^>]+>`)
	doubleAngleRe = regexp.MustCompile(`>>\s*`)
	multiSpaceRe  = regexp.MustCompile(`\s{2,}`)
	wordTimingRe  = regexp.MustCompile(`(\d{2}:\d{2}:\d{2}\.\d{3})><c>([^<]*)</c>`)
	leadTextRe    = regexp.MustCompile(`^([^<]+)<`)
	emotionRe     = regexp.MustCompile(`(?i)\[(cười|thở dài|hắng giọng)\]`)
	bracketRe     = regexp.MustCompile(`\s*\[[^\]]*\]\s*`)
	placeholderRe = regexp.MustCompile("\x00(\\d+)\x00")
)

// symbolsToDrop are read aloud by a synthesiser and mean nothing spoken.
const symbolsToDrop = "♪♫♬→←↑↓↔«»“”‘’„‚"

// parseVTTTime turns "00:01:23.456" into 83.456, or NaN.
func parseVTTTime(raw string) float64 {
	parts := strings.Split(strings.TrimSpace(raw), ":")
	if len(parts) != 3 {
		return math.NaN()
	}
	h, err1 := strconv.Atoi(parts[0])
	m, err2 := strconv.Atoi(parts[1])
	s, err3 := strconv.ParseFloat(parts[2], 64)
	if err1 != nil || err2 != nil || err3 != nil {
		return math.NaN()
	}
	return float64(h)*3600 + float64(m)*60 + s
}

// cleanCueText strips WebVTT markup and the symbols a synthesiser would read
// out. Bracketed descriptions survive this step and are removed by
// stripBrackets, because a bracket can span several <c> tags and is not whole
// yet.
func cleanCueText(raw string) string {
	// Entities first, so &gt;&gt; is >> in time to be recognised below.
	s := strings.NewReplacer(
		"&amp;", "&",
		"&lt;", "<",
		"&gt;", ">",
		"&quot;", `"`,
		"&#39;", "'",
	).Replace(raw)

	s = tagRe.ReplaceAllString(s, "")
	s = doubleAngleRe.ReplaceAllString(s, "")
	s = strings.Map(func(r rune) rune {
		if strings.ContainsRune(symbolsToDrop, r) {
			return -1
		}
		return r
	}, s)
	s = strings.TrimSpace(s)
	return multiSpaceRe.ReplaceAllString(s, " ")
}

// stripBrackets removes [sound effects] and keeps the emotion tags the TTS
// server understands.
func stripBrackets(text string) string {
	var placeholders []string
	s := emotionRe.ReplaceAllStringFunc(text, func(match string) string {
		placeholders = append(placeholders, match)
		return "\x00" + strconv.Itoa(len(placeholders)-1) + "\x00"
	})
	s = bracketRe.ReplaceAllString(s, " ")
	s = strings.ReplaceAll(s, " .", ".")
	s = placeholderRe.ReplaceAllStringFunc(s, func(match string) string {
		i, err := strconv.Atoi(strings.Trim(match, "\x00"))
		if err != nil || i >= len(placeholders) {
			return ""
		}
		return placeholders[i]
	})
	s = multiSpaceRe.ReplaceAllString(s, " ")
	return strings.TrimSpace(s)
}

// piece is one word or phrase with the time it is spoken.
type piece struct {
	start float64
	end   float64
	text  string
}

type rawBlock struct {
	start float64
	end   float64
	lines []string
}

// readBlocks splits the file into timed blocks.
//
// A block's payload ends at a blank line *or* at the next timing line — files in
// the wild have both. Only the blank separator may be stepped over; stepping
// over a timing line discards the cue that line introduced.
func readBlocks(raw string) []rawBlock {
	normalised := strings.ReplaceAll(strings.ReplaceAll(raw, "\r\n", "\n"), "\r", "\n")
	lines := strings.Split(normalised, "\n")

	var blocks []rawBlock
	i := 0
	for i < len(lines) {
		if !strings.Contains(lines[i], "-->") {
			i++
			continue
		}

		timing := lines[i]
		arrow := strings.Index(timing, "-->")
		start := parseVTTTime(timing[:arrow])
		tail := strings.TrimSpace(timing[arrow+3:])
		if cut := strings.IndexAny(tail, " \t"); cut >= 0 {
			tail = tail[:cut]
		}
		end := parseVTTTime(tail)
		i++

		// Leading blank lines belong to the separator, not the payload.
		for i < len(lines) && strings.TrimSpace(lines[i]) == "" {
			i++
		}

		var payload []string
		for i < len(lines) && strings.TrimSpace(lines[i]) != "" &&
			!strings.Contains(lines[i], "-->") {
			payload = append(payload, lines[i])
			i++
		}

		if !math.IsNaN(start) && !math.IsNaN(end) {
			blocks = append(blocks, rawBlock{start: start, end: end, lines: payload})
		}
	}
	return blocks
}

// piecesFromTagged turns one automatic-caption block into word-level pieces.
//
// Only the last payload line is read: in the rolling format the earlier lines
// are the previous cue's text repeated for the viewer, and reading them is what
// makes a phrase say itself twice.
func piecesFromTagged(block rawBlock) []piece {
	line := ""
	if n := len(block.lines); n > 0 {
		line = block.lines[n-1]
	}

	var pieces []piece

	// Text before the first timestamp tag is spoken from the cue's own start.
	if lead := leadTextRe.FindStringSubmatch(line); lead != nil {
		if text := cleanCueText(lead[1]); text != "" {
			pieces = append(pieces, piece{start: block.start, text: text})
		}
	}

	for _, m := range wordTimingRe.FindAllStringSubmatch(line, -1) {
		at := parseVTTTime(m[1])
		text := cleanCueText(m[2])
		if text != "" && !math.IsNaN(at) {
			pieces = append(pieces, piece{start: at, text: text})
		}
	}

	if len(pieces) == 0 {
		if text := cleanCueText(line); text != "" {
			pieces = append(pieces, piece{start: block.start, end: block.end, text: text})
		}
		return pieces
	}

	// Each word runs until the next begins; the last runs to the cue's end.
	for k := range pieces {
		if k+1 < len(pieces) {
			pieces[k].end = pieces[k+1].start
		} else {
			pieces[k].end = block.end
		}
	}
	return pieces
}

// firstClauseBoundary reports the rune offset just past the first clause
// boundary in text, or -1.
//
// The *first*, not the last: taking the last swallows every boundary before it,
// so "Hello there, my friend. How are you" comes out as one cue instead of
// three.
func firstClauseBoundary(text string) int {
	runes := []rune(text)
	for idx, ch := range runes {
		if ch != '.' && ch != '!' && ch != '?' && ch != ',' {
			continue
		}
		var before, after rune
		if idx > 0 {
			before = runes[idx-1]
		}
		if idx+1 < len(runes) {
			after = runes[idx+1]
		}

		// "2.5", "3.14" — a decimal point ends nothing.
		if ch == '.' && isDigit(before) && isDigit(after) {
			continue
		}

		if ch == '.' && isLetter(before) && after == ' ' {
			fields := strings.Fields(string(runes[:idx]))
			if len(fields) > 0 && abbreviations.MatchString(fields[len(fields)-1]) {
				continue
			}
		}

		// Mid-word punctuation is not a boundary.
		if after != 0 && after != ' ' {
			continue
		}

		// A comma is a boundary only when both sides can stand alone. Splitting
		// "Then, how are you" leaves "Then," to be translated with no sentence
		// around it and voiced as a clip half a second long.
		if ch == ',' {
			wordsBefore := len(strings.Fields(string(runes[:idx])))
			wordsAfter := len(strings.Fields(string(runes[idx+1:])))
			if wordsBefore < minClauseWords || wordsAfter < minClauseWords {
				continue
			}
		}

		return idx + 1
	}
	return -1
}

func isDigit(r rune) bool  { return r >= '0' && r <= '9' }
func isLetter(r rune) bool { return (r >= 'a' && r <= 'z') || (r >= 'A' && r <= 'Z') }

// clauseBuilder gathers word pieces into clauses.
//
// A clause carries the start of its first word and the end of its last, which is
// the point of keeping per-word timing this far: a clause claiming the timing of
// the block it came from would be wrong by however much of that block belonged
// to the clause before it.
type clauseBuilder struct {
	buf []piece
	out []vttCue
}

// emit closes a clause at a rune offset into the joined text of the buffer.
func (b *clauseBuilder) emit(upTo int) {
	consumed := 0
	cut := -1
	var leftover *piece

	for k := range b.buf {
		gap := 0
		if k > 0 {
			gap = 1 // the space join inserts
		}
		next := consumed + gap + len([]rune(b.buf[k].text))
		if upTo <= next {
			cut = k
			within := upTo - (consumed + gap)
			runes := []rune(b.buf[k].text)
			if within < 0 {
				within = 0
			}
			if within > len(runes) {
				within = len(runes)
			}
			if rest := strings.TrimSpace(string(runes[within:])); rest != "" {
				leftover = &piece{start: b.buf[k].start, end: b.buf[k].end, text: rest}
			}
			break
		}
		consumed = next
	}
	if cut < 0 {
		cut = len(b.buf) - 1
	}
	if cut < 0 {
		return
	}

	taken := b.buf[:cut+1]
	parts := make([]string, len(taken))
	for i, p := range taken {
		parts[i] = p.text
	}
	joined := []rune(strings.Join(parts, " "))
	if upTo < len(joined) {
		joined = joined[:upTo]
	}
	text := stripBrackets(strings.TrimSpace(string(joined)))
	if text != "" {
		b.out = append(b.out, vttCue{
			Start: taken[0].start,
			End:   taken[len(taken)-1].end,
			Text:  text,
		})
	}

	rest := append([]piece(nil), b.buf[cut+1:]...)
	if leftover != nil {
		b.buf = append([]piece{*leftover}, rest...)
	} else {
		b.buf = rest
	}
}

func (b *clauseBuilder) joined() string {
	parts := make([]string, len(b.buf))
	for i, p := range b.buf {
		parts[i] = p.text
	}
	return strings.Join(parts, " ")
}

func groupIntoClauses(pieces []piece) []vttCue {
	b := &clauseBuilder{}

	for _, p := range pieces {
		b.buf = append(b.buf, p)

		// One block can complete more than one clause.
		for {
			at := firstClauseBoundary(b.joined())
			if at < 0 {
				break
			}
			b.emit(at)
			if len(b.buf) == 0 {
				break
			}
		}

		joined := b.joined()
		if len(strings.Fields(joined)) >= forceSplitWords {
			b.emit(len([]rune(joined)))
		}
	}

	if len(b.buf) > 0 {
		text := stripBrackets(strings.TrimSpace(b.joined()))
		if text != "" {
			b.out = append(b.out, vttCue{
				Start: b.buf[0].start,
				End:   b.buf[len(b.buf)-1].end,
				Text:  text,
			})
		}
	}
	return b.out
}

// parseVTT reads a WebVTT file into the cues the narration speaks.
//
// sourceLang decides whether clause splitting applies: the rules are tuned for
// the punctuation of English and Vietnamese, and a language they do not describe
// is better left in whole cues than cut on rules that do not hold there.
func parseVTT(raw, sourceLang string) []vttCue {
	blocks := readBlocks(raw)

	isAuto := false
	for _, b := range blocks {
		for _, l := range b.lines {
			if strings.Contains(l, "<c>") {
				isAuto = true
				break
			}
		}
		if isAuto {
			break
		}
	}

	var pieces []piece
	var manual []vttCue

	for _, block := range blocks {
		if block.end-block.start < snapshotMaxSeconds {
			continue
		}
		if len(block.lines) == 0 {
			continue
		}

		last := block.lines[len(block.lines)-1]
		if strings.Contains(last, "<c>") {
			pieces = append(pieces, piecesFromTagged(block)...)
			continue
		}

		if isAuto {
			// Rolling text without tags: again only the last line is new.
			if text := cleanCueText(last); text != "" {
				pieces = append(pieces, piece{start: block.start, end: block.end, text: text})
			}
			continue
		}

		if text := cleanCueText(strings.Join(block.lines, " ")); text != "" {
			manual = append(manual, vttCue{Start: block.start, End: block.end, Text: text})
		}
	}

	if !isAuto {
		out := make([]vttCue, 0, len(manual))
		for _, c := range manual {
			if text := stripBrackets(c.Text); text != "" {
				out = append(out, vttCue{Start: c.Start, End: c.End, Text: text})
			}
		}
		return out
	}

	if sourceLang != "en" && sourceLang != "vi" {
		out := make([]vttCue, 0, len(pieces))
		for _, p := range pieces {
			if text := stripBrackets(p.text); text != "" {
				out = append(out, vttCue{Start: p.start, End: p.end, Text: text})
			}
		}
		return out
	}

	return groupIntoClauses(pieces)
}
